File: backend/app/schemas/workspace.py
"""Workspace schemas for API requests/responses"""

from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field


class WorkspaceBase(BaseModel):
    """Base workspace schema"""
    name: str = Field(..., min_length=1, max_length=128, description="Workspace name")
    description: Optional[str] = Field(None, max_length=512, description="Workspace description")


class WorkspaceCreate(WorkspaceBase):
    """Schema for creating a workspace"""
    pass


class WorkspaceUpdate(BaseModel):
    """Schema for updating a workspace"""
    name: Optional[str] = Field(None, min_length=1, max_length=128)
    description: Optional[str] = Field(None, max_length=512)
    is_active: Optional[bool] = None


class WorkspaceOut(WorkspaceBase):
    """Schema for workspace responses"""
    id: int
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class WorkspaceListItem(BaseModel):
    """Simplified workspace info for dropdowns/lists"""
    id: int
    name: str
    is_active: bool

    class Config:
        from_attributes = True


# ===== Multi-Workspace Schemas =====

class UserWorkspaceAssignment(BaseModel):
    """Schema for assigning user to workspace"""
    workspace_id: int
    role: str = Field(default="member", description="User role in workspace")
    is_responsible: bool = Field(default=False, description="Is user responsible for this workspace")

class UserWorkspaceInfo(BaseModel):
    """Schema for user-workspace relationship info"""
    id: int
    name: str
    description: Optional[str]
    role: str
    is_responsible: bool
    assigned_at: Optional[str]

class MeetingWorkspaceAssignment(BaseModel):
    """Schema for assigning meeting to workspace"""
    workspace_id: int
    is_primary: bool = Field(default=False, description="Is this the primary workspace for the meeting")
    relevance_score: int = Field(default=100, ge=0, le=100, description="Relevance score (0-100)")

class MeetingWorkspaceInfo(BaseModel):
    """Schema for meeting-workspace relationship info"""
    id: int
    name: str
    description: Optional[str]
    is_primary: bool
    relevance_score: int
    associated_at: Optional[str]

class MultiWorkspaceAssignment(BaseModel):
    """Schema for assigning to multiple workspaces"""
    workspace_ids: List[int]
    primary_workspace_id: Optional[int] = None

class UserWorkspaceSummary(BaseModel):
    """Summary of user's workspace involvement"""
    user_id: str
    total_workspaces: int
    responsible_workspaces_count: int
    workspaces: List[UserWorkspaceInfo]
    responsible_workspaces: List[UserWorkspaceInfo]

class WorkspaceStats(BaseModel):
    """Workspace statistics"""
    workspace_id: int
    user_count: int
    meeting_count: int
    responsible_user_count: int
    primary_meeting_count: int

class WorkspaceWithStats(WorkspaceOut):
    """Workspace with statistical information"""
    user_count: int = 0
    meeting_count: int = 0
    responsible_user_count: int = 0

File: frontend/src/types/workspace.ts
/**
 * Workspace-related TypeScript interfaces
 * Matches backend Pydantic schemas for type safety
 */

export interface Workspace {
  id: number
  name: string
  description?: string
  is_active: boolean
  created_at: string
  updated_at?: string
}

export interface WorkspaceCreate {
  name: string
  description?: string
}

export interface WorkspaceUpdate {
  name?: string
  description?: string
  is_active?: boolean
}

export interface WorkspaceListItem {
  id: number
  name: string
  is_active: boolean
}

export interface WorkspaceStats {
  workspace: Workspace
  user_count: number
  workspace_meeting_count: number
  personal_meeting_count: number
  total_meetings: number
}

export interface UserWorkspaceAssignment {
  workspace_id?: number | null  // null to remove workspace assignment
}

export interface UserWorkspaceInfo {
  id: string
  username: string
  workspace_id?: number
  workspace_name?: string
  meeting_count: number
  created_at: string
}

// Meeting scope types
export type MeetingScope = 'personal' | 'workspace'

export interface MeetingScopeOption {
  value: MeetingScope
  label: string
  description: string
  disabled?: boolean
}

// API Response types
export interface WorkspaceAssignmentResponse {
  message: string
  user_id: string
  workspace_id?: number
  workspace_name?: string
  previous_workspace?: string
}
